subprojects/workers/src/main/java/org/gradle/workers/internal/DaemonForkOptionsBuilder.java

```diff
@@ -47,7 +56,56 @@ public DaemonForkOptionsBuilder withClassLoaderStructure(ClassLoaderStructure cl
     }
 
     public DaemonForkOptions build() {
-        return new DaemonForkOptions(buildJavaForkOptions(), keepAliveMode, classLoaderStructure);
+        JavaForkOptionsInternal forkOptions = buildJavaForkOptions();
+        if (OperatingSystem.current().isWindows() && keepAliveMode == KeepAliveMode.DAEMON) {
+            List<String> jvmArgs = forkOptions.getAllJvmArgs();
+            if (containsUnreliableOptions(jvmArgs)) {
+                LOGGER.info("Worker requested to be persistent, but its JVM arguments may make the worker unreliable. Worker will expire at the end of the build session.");
+                return new DaemonForkOptions(forkOptions, KeepAliveMode.SESSION, classLoaderStructure);
+            }
+        }
+        return new DaemonForkOptions(forkOptions, keepAliveMode, classLoaderStructure);
```

Comment:
nitpick: I'd extract the "should we change it to session" logic into a method that returns the (possibly changed) keep alive mode and just make this:
```
return new DaemonForkOptions(forkOptions, sessionIfUnreliable(jvmArgs, keepAliveMode), classLoaderStructure);
```

Reply:
I'm not sure about this.  It's a lot easier to see the block above as an exception that only happens in specific cases (Windows, when keep alive is DAEMON, when there are unreliable arguments). If everything is crammed on one line, you have to deal with that complexity every time since the keep alive _might_ change.